reconcile/external_resources/state.py

```diff
@@ -0,0 +1,247 @@
+                            },
+                            self.MODCONF_DRIFT_MINS: {
+                                "N": str(
+                                    state.reconciliation.module_configuration.reconcile_drift_interval_minutes
+                                )
+                            },
+                            self.MODCONF_TIMEOUT_MINS: {
+                                "N": str(
+                                    state.reconciliation.module_configuration.reconcile_timeout_minutes
+                                )
+                            },
+                        }
+                    },
+                }
+            },
+        }
+
+
+class ExternalResourcesStateDynamoDB:
+    PARTIALS_PROJECTED_VALUES = ",".join([
+        DynamoDBStateAdapter.ER_KEY,
+        DynamoDBStateAdapter.TIMESTAMP,
+        DynamoDBStateAdapter.RESOURCE_STATUS,
+        DynamoDBStateAdapter.RECONCILIATION_ERRORS,
+        f"{DynamoDBStateAdapter.RECONC}.{DynamoDBStateAdapter.RECONC_RESOURCE_HASH}",
+    ])
+
+    def __init__(self, table_name: str, index_name: str, region_name: str) -> None:
+        self.adapter = DynamoDBStateAdapter()
+        self.client = boto3.client("dynamodb", region_name=region_name)
+        self._table = table_name
+        self._index_name = index_name
+        self.partial_resources = self._get_partial_resources()
+
+    def get_external_resource_state(
+        self, key: ExternalResourceKey
+    ) -> ExternalResourceState:
+        data = self.client.get_item(
+            TableName=self._table,
+            ConsistentRead=True,
+            Key={self.adapter.ER_KEY_HASH: {"S": key.hash()}},
+        )
+        if "Item" in data:
+            return self.adapter.deserialize(data["Item"])
+        else:
+            return ExternalResourceState(
+                key=key,
+                ts=datetime.now(timezone.utc),
+                resource_status=ResourceStatus.NOT_EXISTS,
+                reconciliation=Reconciliation(key=key),
+                reconciliation_errors=0,
+            )
+
+    def set_external_resource_state(
+        self,
+        state: ExternalResourceState,
+    ) -> None:
+        self.client.put_item(TableName=self._table, Item=self.adapter.serialize(state))
+
+    def del_external_resource_state(self, key: ExternalResourceKey) -> None:
+        self.client.delete_item(
+            TableName=self._table,
+            Key={self.adapter.ER_KEY_HASH: {"S": key.hash()}},
+        )
+
+    def _get_partial_resources(
+        self,
+    ) -> dict[ExternalResourceKey, ExternalResourceState]:
+        """A Partial Resoure is the minimum resource data reguired
+        to check if a resource has been removed from the configuration.
+        Getting less data from DynamoDb saves money and the logic does not need it.
+        """
+        logging.info("Getting Managed resources from DynamoDb")
+        partials = {}
+        for item in self.client.scan(
```

Comment:
I think this is the paginator https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/dynamodb/paginator/Scan.html